{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "d68b1267-009c-41b6-8ea3-68d65833a886",
   "metadata": {},
   "source": [
    "## 25th June 2025\n",
    "## TCGA Data Folder Exploration \n",
    "## "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "008044a7-efef-47b8-ab3e-ecbe1066abd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#install libraries\n",
    "#!pip install pandas scikit-learn numpy matplot"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "012ba1b0-abed-4468-9a5a-b24c550d9445",
   "metadata": {},
   "outputs": [],
   "source": [
    "#import libraries\n",
    "#data handling\n",
    "import pandas as pd #to manipulate data\n",
    "import os #to manipulate working directory\n",
    "import numpy as np\n",
    "\n",
    "#data visualization\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "#preprocessing\n",
    "from sklearn.model_selection import train_test_split\n",
    "from sklearn.preprocessing import StandardScaler\n",
    "from sklearn.preprocessing import LabelEncoder\n",
    "from sklearn.preprocessing import label_binarize\n",
    "\n",
    "#Feature selection\n",
    "from sklearn.feature_selection import mutual_info_classif\n",
    "\n",
    "#Classification\n",
    "from sklearn.multiclass import OneVsRestClassifier\n",
    "from sklearn.ensemble import RandomForestClassifier\n",
    "\n",
    "#performance metrics \n",
    "from sklearn.metrics import balanced_accuracy_score,f1_score,precision_score, recall_score\n",
    "from sklearn.metrics import roc_auc_score\n",
    "from sklearn.metrics import roc_curve,auc\n",
    "from sklearn.metrics import classification_report\n",
    "from sklearn.metrics import confusion_matrix\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c72845a0-aad1-4fb9-add8-a98fc7e0d292",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Current working directory: C:\\Users\\peace\\OneDrive\\Desktop\\Project\\DataFile\\prad_tcga_pan_can_atlas_2018\\prad_tcga_pan_can_atlas_2018\n"
     ]
    }
   ],
   "source": [
    "# Read the Data\n",
    "folder_path = r'C:\\Users\\peace\\OneDrive\\Desktop\\Project\\DataFile\\prad_tcga_pan_can_atlas_2018\\prad_tcga_pan_can_atlas_2018'\n",
    "# Change the working directory\n",
    "os.chdir(folder_path)\n",
    "\n",
    "# Check the working directory\n",
    "print(\"Current working directory:\", os.getcwd())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "93dbae40-cceb-4f08-8b84-7d540cb128e2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Files found: ['data_armlevel_cna.txt', 'data_clinical_patient.txt', 'data_clinical_sample.txt', 'data_clinical_supp_hypoxia.txt', 'data_cna.txt', 'data_genetic_ancestry.txt', 'data_gene_panel_matrix.txt', 'data_log2_cna.txt', 'data_methylation_hm27_hm450_merged.txt', 'data_mrna_seq_v2_rsem.txt', 'data_mrna_seq_v2_rsem_zscores_ref_all_samples.txt', 'data_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt', 'data_mrna_seq_v2_rsem_zscores_ref_normal_samples.txt', 'data_mutations.txt', 'data_resource_definition.txt', 'data_resource_patient.txt', 'data_rppa.txt', 'data_rppa_zscores.txt', 'data_sv.txt', 'data_timeline_sample_acquisition.txt', 'data_timeline_status.txt', 'data_timeline_treatment.txt', 'meta_armlevel_cna.txt', 'meta_clinical_patient.txt', 'meta_clinical_sample.txt', 'meta_cna.txt', 'meta_cna_hg19_seg.txt', 'meta_genetic_ancestry.txt', 'meta_gene_panel_matrix.txt', 'meta_log2_cna.txt', 'meta_methylation_hm27_hm450_merged.txt', 'meta_mrna_seq_v2_rsem.txt', 'meta_mrna_seq_v2_rsem_zscores_ref_all_samples.txt', 'meta_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt', 'meta_mrna_seq_v2_rsem_zscores_ref_normal_samples.txt', 'meta_mutations.txt', 'meta_resource_definition.txt', 'meta_resource_patient.txt', 'meta_rppa.txt', 'meta_rppa_zscores.txt', 'meta_study.txt', 'meta_sv.txt', 'meta_timeline_sample_acquisition.txt', 'meta_timeline_status.txt', 'meta_timeline_treatment.txt']\n",
      "data_armlevel_cna.txt loaded successfully. Shape: (39, 497)\n",
      "data_clinical_patient.txt loaded successfully. Shape: (498, 38)\n",
      "data_clinical_sample.txt loaded successfully. Shape: (498, 19)\n",
      "data_clinical_supp_hypoxia.txt loaded successfully. Shape: (338, 4)\n",
      "data_cna.txt loaded successfully. Shape: (25128, 491)\n",
      "data_genetic_ancestry.txt loaded successfully. Shape: (5, 468)\n",
      "data_gene_panel_matrix.txt loaded successfully. Shape: (494, 4)\n",
      "data_log2_cna.txt loaded successfully. Shape: (25128, 491)\n",
      "data_methylation_hm27_hm450_merged.txt loaded successfully. Shape: (22601, 498)\n",
      "data_mrna_seq_v2_rsem.txt loaded successfully. Shape: (20531, 495)\n",
      "data_mrna_seq_v2_rsem_zscores_ref_all_samples.txt loaded successfully. Shape: (20531, 495)\n",
      "data_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt loaded successfully. Shape: (20471, 495)\n",
      "data_mrna_seq_v2_rsem_zscores_ref_normal_samples.txt loaded successfully. Shape: (20531, 495)\n",
      "data_mutations.txt loaded successfully. Shape: (34192, 114)\n",
      "data_resource_definition.txt loaded successfully. Shape: (1, 6)\n",
      "data_resource_patient.txt loaded successfully. Shape: (500, 3)\n",
      "data_rppa.txt loaded successfully. Shape: (198, 351)\n",
      "data_rppa_zscores.txt loaded successfully. Shape: (198, 351)\n",
      "data_sv.txt loaded successfully. Shape: (1891, 13)\n",
      "data_timeline_sample_acquisition.txt loaded successfully. Shape: (485, 14)\n",
      "data_timeline_status.txt loaded successfully. Shape: (1569, 20)\n",
      "data_timeline_treatment.txt loaded successfully. Shape: (208, 31)\n",
      "meta_armlevel_cna.txt loaded successfully. Shape: (9, 1)\n",
      "meta_clinical_patient.txt loaded successfully. Shape: (3, 1)\n",
      "meta_clinical_sample.txt loaded successfully. Shape: (3, 1)\n",
      "meta_cna.txt loaded successfully. Shape: (7, 1)\n",
      "meta_cna_hg19_seg.txt loaded successfully. Shape: (8, 1)\n",
      "meta_genetic_ancestry.txt loaded successfully. Shape: (10, 1)\n",
      "meta_gene_panel_matrix.txt loaded successfully. Shape: (3, 1)\n",
      "meta_log2_cna.txt loaded successfully. Shape: (7, 1)\n",
      "meta_methylation_hm27_hm450_merged.txt loaded successfully. Shape: (10, 1)\n",
      "meta_mrna_seq_v2_rsem.txt loaded successfully. Shape: (7, 1)\n",
      "meta_mrna_seq_v2_rsem_zscores_ref_all_samples.txt loaded successfully. Shape: (7, 1)\n",
      "meta_mrna_seq_v2_rsem_zscores_ref_diploid_samples.txt loaded successfully. Shape: (7, 1)\n",
      "meta_mrna_seq_v2_rsem_zscores_ref_normal_samples.txt loaded successfully. Shape: (7, 1)\n",
      "meta_mutations.txt loaded successfully. Shape: (7, 1)\n",
      "meta_resource_definition.txt loaded successfully. Shape: (2, 1)\n",
      "meta_resource_patient.txt loaded successfully. Shape: (2, 1)\n",
      "meta_rppa.txt loaded successfully. Shape: (7, 1)\n",
      "meta_rppa_zscores.txt loaded successfully. Shape: (7, 1)\n",
      "meta_study.txt loaded successfully. Shape: (7, 1)\n",
      "meta_sv.txt loaded successfully. Shape: (7, 1)\n",
      "meta_timeline_sample_acquisition.txt loaded successfully. Shape: (3, 1)\n",
      "meta_timeline_status.txt loaded successfully. Shape: (3, 1)\n",
      "meta_timeline_treatment.txt loaded successfully. Shape: (3, 1)\n"
     ]
    }
   ],
   "source": [
    "txt_files = [f for f in os.listdir() if f.endswith('.txt')]   #iterates through the file list for files that ends with \".txt\"\n",
    "print(\"Files found:\", txt_files)\n",
    "\n",
    "# Create an empty Dictionary to store the file as DataFrames\n",
    "dataframes = {}\n",
    "\n",
    "# Load each file\n",
    "for file in txt_files:              #iterate through each \".txt\" file \n",
    "    try:                            #use try to load file to avoid error into dataframe \n",
    "        df = pd.read_csv(file, sep='\\t', low_memory=False)   #load into panda df\n",
    "        key_name = os.path.splitext(file)[0]  #split filename \"cna.txt\" > \"cna\" \"txt\" & select the first part[0] to assign as key name \n",
    "        dataframes[key_name] = df             #store df into the dictionary using file name as key\n",
    "        print(f\"{file} loaded successfully. Shape: {df.shape}\")\n",
    "    except Exception as e:                    \n",
    "        print(f\"Failed to load {file}: {e}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "d589d91c-fb55-4848-90d3-21eddfb10a88",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (20531, 495)\n",
      "First 10 column names:\n",
      " ['Hugo_Symbol', 'Entrez_Gene_Id', 'TCGA-2A-A8VL-01', 'TCGA-2A-A8VO-01', 'TCGA-2A-A8VT-01', 'TCGA-2A-A8VV-01', 'TCGA-2A-A8VX-01', 'TCGA-2A-A8W1-01', 'TCGA-2A-A8W3-01', 'TCGA-2A-AAYF-01']\n",
      "Shape: (25128, 491)\n",
      "First 10 column names:\n",
      " ['Hugo_Symbol', 'Entrez_Gene_Id', 'TCGA-2A-A8VL-01', 'TCGA-2A-A8VO-01', 'TCGA-2A-A8VT-01', 'TCGA-2A-A8VV-01', 'TCGA-2A-A8VX-01', 'TCGA-2A-A8W1-01', 'TCGA-2A-A8W3-01', 'TCGA-2A-AAYF-01']\n",
      "['PATIENT_ID', 'START_DATE', 'STOP_DATE', 'EVENT_TYPE', 'TREATMENT_TYPE', 'TREATMENT_SUBTYPE', 'AGENT', 'NUMBER_OF_CYCLES', 'PRESCRIBED_DOSE', 'PRESCRIBED_DOSE_UNITS', 'REGIMEN_NUMBER', 'REGIMEN_INDICATION', 'REGIMEN_INDICATION_NOTES', 'MEASURE_OF_RESPONSE', 'CLINICAL_TRIAL_DRUG_CLASSIFICATION', 'ROUTE_OF_ADMINISTRATION', 'ROUTE_OF_ADMINISTRATION-2', 'THERAPY_ONGOING', 'TOTAL_DOSE', 'TOTAL_DOSE_UNITS', 'TX_ON_CLINICAL_TRIAL', 'ANATOMIC_TREATMENT_SITE', 'COURSE_NUMBER', 'NUMBER_OF_FRACTIONS', 'RADIATION_DOSAGE', 'RADIATION_TREATMENT_ONGOING', 'RADIATION_TYPE', 'RADIATION_UNITS', 'PHARM_REGIMEN', 'STEM_CELL_TRANSPLANTATION', 'STEM_CELL_TRANSPLANTATION_TYPE']\n"
     ]
    }
   ],
   "source": [
    "#GENE and Treatment Identification\n",
    "\n",
    "#Z-score normalized RNA-Seq data showing gene expression levels across patient samples,\n",
    "#used to analyze MAPK pathway activity and treatment response.\n",
    "mrna_df = dataframes['data_mrna_seq_v2_rsem_zscores_ref_all_samples']\n",
    "print(\"Shape:\", mrna_df.shape)\n",
    "print(\"First 10 column names:\\n\", mrna_df.columns.tolist()[:10])\n",
    "\n",
    "#Load and assess Copy Number Alteration (CNA) data \n",
    "cna_df = dataframes['data_log2_cna']\n",
    "print(\"Shape:\", cna_df.shape)\n",
    "print(\"First 10 column names:\\n\", cna_df.columns.tolist()[:10])\n",
    "\n",
    "#timeline treatment\n",
    "treatment_df = dataframes['data_timeline_treatment']\n",
    "print(treatment_df.columns.tolist())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "0fd13519-b1b7-4fa7-8877-a890765bdae9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MEASURE_OF_RESPONSE\n",
      "NaN                                 77\n",
      "complete response                   76\n",
      "partial response                    22\n",
      "stable disease                      16\n",
      "clinical progressive disease        10\n",
      "radiographic progressive disease     7\n",
      "Name: count, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Measure of treatment response is the target variable\n",
    "print(treatment_df['MEASURE_OF_RESPONSE'].value_counts(dropna=False))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "6daece9a-3cb3-446c-9d20-e88e2cd8d6e6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Transposed mRNA shape: (494, 20533)\n",
      "Transposed CNA shape: (490, 25130)\n"
     ]
    }
   ],
   "source": [
    "## Rearranging the data \n",
    "#Setting gene names as index and transpose \n",
    "\n",
    "mrna_trans_df = mrna_df\n",
    "\n",
    "#Transpose the mRNA data\n",
    "mrna_trans_df = mrna_df.set_index(mrna_df.columns[0]).T    #set index column as the new index column and transpose\n",
    "\n",
    "#Reset the index to get the Sample_ID\n",
    "mrna_trans_df.reset_index(inplace= True)             #moves the index to a new column \n",
    "mrna_trans_df.rename(columns= {'index': 'Sample_ID'}, inplace=True)  #rename the index as 'sample_ID' for clarity \n",
    "\n",
    "#Trimming sample IDs to first 12 chars to match patient-level ID format\n",
    "mrna_trans_df['PATIENT_ID'] = mrna_trans_df['Sample_ID'].str[:12]      #Take the first 12 characters of the sample_ID to create a 'PATIENT_ID' column\n",
    "\n",
    "print(\"Transposed mRNA shape:\",mrna_trans_df.shape)\n",
    "\n",
    "\n",
    "#set gene names as index and transpose the CNA data \n",
    "cna_trans_df =cna_df \n",
    "\n",
    "#Transpose the CNA Ddata\n",
    "cna_trans_df = cna_df.set_index(cna_df.columns[0]).T\n",
    "\n",
    "#Reset index to get Sample_ID\n",
    "cna_trans_df.reset_index(inplace= True)\n",
    "cna_trans_df.rename(columns={'index': 'Sample_ID'}, inplace=True)\n",
    "\n",
    "#Triming sample IDs to first 12 chars to match patient-level ID format\n",
    "cna_trans_df['PATIENT_ID'] = cna_trans_df['Sample_ID'].str[:12]\n",
    "\n",
    "print(\"Transposed CNA shape:\", cna_trans_df.shape)\n",
    "\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e50c1d4c-fae9-42b9-adeb-ed5ee0d03e45",
   "metadata": {},
   "source": [
    "### Preparing the MAPK Gene List\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "4e768f53-4446-40f0-81c9-457732adf4eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAPK genes found in data: ['ARAF', 'BRAF', 'HRAS', 'KRAS', 'NRAS', 'RAF1', 'MAP2K1', 'MAP2K2', 'MAPK1', 'MAPK3', 'MAPK8', 'MAPK9', 'MAPK10', 'MAPK11', 'MAPK12', 'MAPK13', 'MAPK14', 'MAPK15', 'DUSP1', 'PTEN', 'EGFR', 'KSR1', 'GRB2', 'SOS1', 'FOS', 'JUN', 'MAP3K1', 'MAP3K2', 'DUSP4', 'DUSP6', 'RASA1', 'SPRY2', 'FGFR1', 'FGFR2']\n",
      "MAPK genes missing from data: []\n"
     ]
    }
   ],
   "source": [
    "mapk_genes = [\n",
    "    'ARAF', 'BRAF', 'HRAS', 'KRAS', 'NRAS', 'RAF1',\n",
    "    'MAP2K1', 'MAP2K2', 'MAPK1', 'MAPK3',\n",
    "    'MAPK8', 'MAPK9', 'MAPK10', 'MAPK11', 'MAPK12',\n",
    "    'MAPK13', 'MAPK14', 'MAPK15',\n",
    "    'DUSP1', 'PTEN', 'EGFR', 'KSR1', 'GRB2', 'SOS1', 'FOS', 'JUN', 'MAP3K1', 'MAP3K2',\n",
    "    'DUSP4', 'DUSP6', 'RASA1', 'SPRY2', 'FGFR1', 'FGFR2'\n",
    "]  # there are 34 MAPK genes\n",
    "\n",
    "#checking for absent or present gene in the MAPK dataset\n",
    "present_genes = [gene for gene in mapk_genes if gene in mrna_trans_df.columns]\n",
    "missing_genes = [gene for gene in mapk_genes if gene not in mrna_trans_df.columns]\n",
    "\n",
    "print(\"MAPK genes found in data:\", present_genes)\n",
    "print(\"MAPK genes missing from data:\", missing_genes)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c411655f-b69b-4e4f-b5c8-ee78bc75cfd1",
   "metadata": {},
   "source": [
    "The result shows that all 34 MAPK pathway genes are present in the mRNA dataset. This confirms that the dataset includes expression profiles for key genes involved in the MAPK signaling pathway, which are relevant for studying treatment resistance in prostate cancer. With these genes available, they can now be extracted as input features for training the machine learning model. The analysis will have biologically significant biomarkers (MAPK) associated with treatment outcomes."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a920a01d-b979-4a0b-abc8-bddae0501858",
   "metadata": {},
   "source": [
    "### Preparing the MAPK Gene Data Set\n",
    "Selected the MAPK genes that will be used as the train features for the MAPK genes. The dataset from the mrna data and then the cna data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "f51267d0-e1d6-4393-afa6-00016c581511",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MAPK gene expression data shape: (494, 35)\n",
      "Hugo_Symbol    PATIENT_ID      ARAF      BRAF       HRAS       KRAS  \\\n",
      "0            Entrez_Gene_  369.0000  673.0000  3265.0000  3845.0000   \n",
      "1            TCGA-2A-A8VL   -0.0610   -1.5311     0.2898    -1.0884   \n",
      "2            TCGA-2A-A8VO    0.8460   -1.1840     0.7671    -0.3870   \n",
      "3            TCGA-2A-A8VT   -0.2260    1.1105     0.1345    -1.3060   \n",
      "4            TCGA-2A-A8VV   -1.0212    0.1110    -0.0572    -0.5366   \n",
      "\n",
      "Hugo_Symbol       NRAS       RAF1     MAP2K1     MAP2K2      MAPK1  ...  \\\n",
      "0            4893.0000  5894.0000  5604.0000  5605.0000  5594.0000  ...   \n",
      "1              -1.0793    -0.3868     0.9475     0.9211     0.0663  ...   \n",
      "2              -0.0213    -1.0856     0.7794     1.1497    -0.5735  ...   \n",
      "3               0.4124     2.8254     1.2351    -0.4458    -0.4918  ...   \n",
      "4              -0.8571     0.0795     0.5959     0.5107    -0.2450  ...   \n",
      "\n",
      "Hugo_Symbol        FOS        JUN     MAP3K1      MAP3K2      DUSP4  \\\n",
      "0            2353.0000  3725.0000  4214.0000  10746.0000  1846.0000   \n",
      "1               1.3174     1.3805    -0.7783     -1.1798    -0.4250   \n",
      "2              -0.1463     0.6179    -0.6543     -0.5415    -1.7709   \n",
      "3              -0.2025     0.0554     0.7225      1.6066    -1.8192   \n",
      "4              -1.6768    -0.7916    -0.5662     -0.6408    -0.2874   \n",
      "\n",
      "Hugo_Symbol      DUSP6      RASA1       SPRY2      FGFR1      FGFR2  \n",
      "0            1848.0000  5921.0000  10253.0000  2260.0000  2263.0000  \n",
      "1               0.6427    -0.3476      0.0091    -0.5909     0.4665  \n",
      "2              -0.3789    -0.5374     -0.0030    -0.2105    -1.6122  \n",
      "3              -1.0859     0.5282      0.4044    -1.8975     0.1760  \n",
      "4              -0.9959     0.0238     -0.3172     0.5303    -0.3753  \n",
      "\n",
      "[5 rows x 35 columns]\n",
      "\n",
      " MAPK gene expression data shape: (490, 35)\n",
      "Hugo_Symbol    PATIENT_ID  ARAF_cna  BRAF_cna  HRAS_cna  KRAS_cna  NRAS_cna  \\\n",
      "0            Entrez_Gene_   369.000   673.000  3265.000  3845.000  4893.000   \n",
      "1            TCGA-2A-A8VL    -0.016     0.001     0.010     0.001    -0.485   \n",
      "2            TCGA-2A-A8VO    -0.015     0.002     0.000    -0.002     0.002   \n",
      "3            TCGA-2A-A8VT    -0.097     0.061     0.076    -0.690     0.046   \n",
      "4            TCGA-2A-A8VV    -0.003     0.002     0.000     0.000    -0.001   \n",
      "\n",
      "Hugo_Symbol  RAF1_cna  MAP2K1_cna  MAP2K2_cna  MAPK1_cna  ...   FOS_cna  \\\n",
      "0            5894.000    5604.000    5605.000   5594.000  ...  2353.000   \n",
      "1              -0.025       0.005       0.010      0.012  ...    -0.000   \n",
      "2               0.005       0.004       0.004      0.009  ...     0.003   \n",
      "3               0.059      -0.025       0.013     -0.369  ...    -0.046   \n",
      "4               0.004       0.003       0.009      0.010  ...    -0.023   \n",
      "\n",
      "Hugo_Symbol   JUN_cna  MAP3K1_cna  MAP3K2_cna  DUSP4_cna  DUSP6_cna  \\\n",
      "0            3725.000    4214.000   10746.000   1846.000   1848.000   \n",
      "1               0.000      -0.002      -0.002     -0.002     -0.003   \n",
      "2               0.002      -0.001       0.002     -0.572     -0.002   \n",
      "3               0.046      -0.285       0.036     -1.070      0.000   \n",
      "4              -0.545      -0.002      -0.016     -0.214      0.003   \n",
      "\n",
      "Hugo_Symbol  RASA1_cna  SPRY2_cna  FGFR1_cna  FGFR2_cna  \n",
      "0             5921.000  10253.000   2260.000   2263.000  \n",
      "1               -0.002     -0.009     -0.002      0.001  \n",
      "2               -0.001     -0.019     -0.002     -0.000  \n",
      "3               -0.285      0.070     -1.070     -0.247  \n",
      "4               -0.002     -0.013     -0.214      0.002  \n",
      "\n",
      "[5 rows x 35 columns]\n"
     ]
    }
   ],
   "source": [
    "mapk_mrna_df = mrna_trans_df[['PATIENT_ID'] + mapk_genes]\n",
    "print(\"MAPK gene expression data shape:\", mapk_mrna_df.shape)\n",
    "print(mapk_mrna_df.head())\n",
    "\n",
    "#The cna data \n",
    "mapk_cna_df = cna_trans_df[['PATIENT_ID'] + mapk_genes]\n",
    "mapk_cna_df = mapk_cna_df.rename(columns = {gene: f\"{gene}_cna\" for gene in mapk_genes})\n",
    "print(\"\\n MAPK gene expression data shape:\", mapk_cna_df.shape)\n",
    "print(mapk_cna_df.head())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ade64ba-f358-470c-a7df-a8cd39443420",
   "metadata": {},
   "source": [
    "### Prepare the Target Dataset\n",
    "Treatment was set as  the target "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "e659726d-97fc-46db-840c-77e588634693",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['PATIENT_ID', 'MEASURE_OF_RESPONSE', 'TREATMENT_TYPE']\n",
      "target dataframe shape: (208, 3)\n"
     ]
    }
   ],
   "source": [
    "#\"Patiient id\" is the primary key for the table. \"Measure of response\" is the target variable\n",
    "target_df = treatment_df[['PATIENT_ID','MEASURE_OF_RESPONSE','TREATMENT_TYPE']]\n",
    "\n",
    "print(target_df.columns.tolist())\n",
    "print(\"target dataframe shape:\", target_df.shape)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8322da70-1ffc-4416-bab9-4b6c25ef2c1f",
   "metadata": {},
   "source": [
    "### Join MAPK Gene and Response  \n",
    "To prepare the dataset for machine learning, a join operation was used between the MAPK gene expression data and the treatment response labels. Both datasets contain a common identifier \"PATIENT_ID2,which was used as the primary key for merging both data.\n",
    "An inner join was applied to retain only patients who had both gene expression profiles and recorded tretament outcomes. This stepp was used to ensure the resulting dataset is clean and well-structured, contaonong only complete cases suitable for suppervised learning and prsictive model."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "09f00b46-8ed0-436d-8e54-cec66b57c155",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Final merged DataFrame shape: (208, 71)\n",
      "     PATIENT_ID    ARAF    BRAF    HRAS    KRAS    NRAS    RAF1  MAP2K1  \\\n",
      "0  TCGA-2A-A8VT -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351   \n",
      "1  TCGA-2A-A8VT -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351   \n",
      "2  TCGA-4L-AA1F -1.2337  0.0838 -0.6688  0.6233  1.0952 -0.9766 -1.0624   \n",
      "3  TCGA-CH-5741 -0.0553 -0.1687 -0.5289 -0.0316 -0.0179  0.8442  0.9456   \n",
      "4  TCGA-CH-5751 -1.5708  3.5178 -2.1029  2.1736  2.3636  1.1228 -0.4158   \n",
      "\n",
      "   MAP2K2   MAPK1  ...  MAP3K1_cna  MAP3K2_cna  DUSP4_cna  DUSP6_cna  \\\n",
      "0 -0.4458 -0.4918  ...      -0.285       0.036     -1.070      0.000   \n",
      "1 -0.4458 -0.4918  ...      -0.285       0.036     -1.070      0.000   \n",
      "2 -1.0342  0.1173  ...      -0.144       0.084     -0.256      0.072   \n",
      "3 -0.8248 -0.8615  ...       0.000      -0.006      0.004     -0.002   \n",
      "4 -1.4598  0.5000  ...      -0.731      -0.015     -0.668     -0.203   \n",
      "\n",
      "   RASA1_cna  SPRY2_cna  FGFR1_cna  FGFR2_cna  MEASURE_OF_RESPONSE  \\\n",
      "0     -0.285      0.070     -1.070     -0.247       stable disease   \n",
      "1     -0.285      0.070     -1.070     -0.247       stable disease   \n",
      "2     -0.144     -0.308     -0.256     -0.238                  NaN   \n",
      "3      0.000     -0.701      0.004     -0.000    complete response   \n",
      "4     -0.111     -0.425      0.844     -0.151                  NaN   \n",
      "\n",
      "      TREATMENT_TYPE  \n",
      "0    Hormone Therapy  \n",
      "1  Radiation Therapy  \n",
      "2    Hormone Therapy  \n",
      "3    Hormone Therapy  \n",
      "4    Hormone Therapy  \n",
      "\n",
      "[5 rows x 71 columns]\n"
     ]
    }
   ],
   "source": [
    "#Merging the mRNA (mapk_mrna_df) and CNA (mapk_cna_df) data \n",
    "mapk_combined = mapk_mrna_df.merge(mapk_cna_df, on='PATIENT_ID', how='inner')\n",
    "\n",
    "# Merging with response variable (target_df)\n",
    "merged_df = mapk_combined.merge(target_df, on='PATIENT_ID', how='inner')\n",
    "\n",
    "#View the merged data sets \n",
    "print(\"Final merged DataFrame shape:\", merged_df.shape)\n",
    "print(merged_df.head())\n",
    "\n",
    "#Export\n",
    "merged_df.to_csv('merged_df.csv', index= False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4d72ffcd-b659-4b3e-9f2b-de6c94237304",
   "metadata": {},
   "source": [
    "### Importing Dataset to Environment from the Working Directory\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "619586df-eadf-4a31-a24f-50407f70b0d7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Merged DataFrame shape: (208, 71)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>PATIENT_ID</th>\n",
       "      <th>ARAF</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>HRAS</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>NRAS</th>\n",
       "      <th>RAF1</th>\n",
       "      <th>MAP2K1</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MAPK1</th>\n",
       "      <th>...</th>\n",
       "      <th>MAP3K1_cna</th>\n",
       "      <th>MAP3K2_cna</th>\n",
       "      <th>DUSP4_cna</th>\n",
       "      <th>DUSP6_cna</th>\n",
       "      <th>RASA1_cna</th>\n",
       "      <th>SPRY2_cna</th>\n",
       "      <th>FGFR1_cna</th>\n",
       "      <th>FGFR2_cna</th>\n",
       "      <th>MEASURE_OF_RESPONSE</th>\n",
       "      <th>TREATMENT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>TCGA-2A-A8VT</td>\n",
       "      <td>-0.2260</td>\n",
       "      <td>1.1105</td>\n",
       "      <td>0.1345</td>\n",
       "      <td>-1.3060</td>\n",
       "      <td>0.4124</td>\n",
       "      <td>2.8254</td>\n",
       "      <td>1.2351</td>\n",
       "      <td>-0.4458</td>\n",
       "      <td>-0.4918</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.036</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>stable disease</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TCGA-2A-A8VT</td>\n",
       "      <td>-0.2260</td>\n",
       "      <td>1.1105</td>\n",
       "      <td>0.1345</td>\n",
       "      <td>-1.3060</td>\n",
       "      <td>0.4124</td>\n",
       "      <td>2.8254</td>\n",
       "      <td>1.2351</td>\n",
       "      <td>-0.4458</td>\n",
       "      <td>-0.4918</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.036</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>stable disease</td>\n",
       "      <td>Radiation Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>TCGA-4L-AA1F</td>\n",
       "      <td>-1.2337</td>\n",
       "      <td>0.0838</td>\n",
       "      <td>-0.6688</td>\n",
       "      <td>0.6233</td>\n",
       "      <td>1.0952</td>\n",
       "      <td>-0.9766</td>\n",
       "      <td>-1.0624</td>\n",
       "      <td>-1.0342</td>\n",
       "      <td>0.1173</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.144</td>\n",
       "      <td>0.084</td>\n",
       "      <td>-0.256</td>\n",
       "      <td>0.072</td>\n",
       "      <td>-0.144</td>\n",
       "      <td>-0.308</td>\n",
       "      <td>-0.256</td>\n",
       "      <td>-0.238</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>TCGA-CH-5741</td>\n",
       "      <td>-0.0553</td>\n",
       "      <td>-0.1687</td>\n",
       "      <td>-0.5289</td>\n",
       "      <td>-0.0316</td>\n",
       "      <td>-0.0179</td>\n",
       "      <td>0.8442</td>\n",
       "      <td>0.9456</td>\n",
       "      <td>-0.8248</td>\n",
       "      <td>-0.8615</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.006</td>\n",
       "      <td>0.004</td>\n",
       "      <td>-0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.701</td>\n",
       "      <td>0.004</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>complete response</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TCGA-CH-5751</td>\n",
       "      <td>-1.5708</td>\n",
       "      <td>3.5178</td>\n",
       "      <td>-2.1029</td>\n",
       "      <td>2.1736</td>\n",
       "      <td>2.3636</td>\n",
       "      <td>1.1228</td>\n",
       "      <td>-0.4158</td>\n",
       "      <td>-1.4598</td>\n",
       "      <td>0.5000</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.731</td>\n",
       "      <td>-0.015</td>\n",
       "      <td>-0.668</td>\n",
       "      <td>-0.203</td>\n",
       "      <td>-0.111</td>\n",
       "      <td>-0.425</td>\n",
       "      <td>0.844</td>\n",
       "      <td>-0.151</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 71 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     PATIENT_ID    ARAF    BRAF    HRAS    KRAS    NRAS    RAF1  MAP2K1  \\\n",
       "0  TCGA-2A-A8VT -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351   \n",
       "1  TCGA-2A-A8VT -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351   \n",
       "2  TCGA-4L-AA1F -1.2337  0.0838 -0.6688  0.6233  1.0952 -0.9766 -1.0624   \n",
       "3  TCGA-CH-5741 -0.0553 -0.1687 -0.5289 -0.0316 -0.0179  0.8442  0.9456   \n",
       "4  TCGA-CH-5751 -1.5708  3.5178 -2.1029  2.1736  2.3636  1.1228 -0.4158   \n",
       "\n",
       "   MAP2K2   MAPK1  ...  MAP3K1_cna  MAP3K2_cna  DUSP4_cna  DUSP6_cna  \\\n",
       "0 -0.4458 -0.4918  ...      -0.285       0.036     -1.070      0.000   \n",
       "1 -0.4458 -0.4918  ...      -0.285       0.036     -1.070      0.000   \n",
       "2 -1.0342  0.1173  ...      -0.144       0.084     -0.256      0.072   \n",
       "3 -0.8248 -0.8615  ...       0.000      -0.006      0.004     -0.002   \n",
       "4 -1.4598  0.5000  ...      -0.731      -0.015     -0.668     -0.203   \n",
       "\n",
       "   RASA1_cna  SPRY2_cna  FGFR1_cna  FGFR2_cna  MEASURE_OF_RESPONSE  \\\n",
       "0     -0.285      0.070     -1.070     -0.247       stable disease   \n",
       "1     -0.285      0.070     -1.070     -0.247       stable disease   \n",
       "2     -0.144     -0.308     -0.256     -0.238                  NaN   \n",
       "3      0.000     -0.701      0.004     -0.000    complete response   \n",
       "4     -0.111     -0.425      0.844     -0.151                  NaN   \n",
       "\n",
       "      TREATMENT_TYPE  \n",
       "0    Hormone Therapy  \n",
       "1  Radiation Therapy  \n",
       "2    Hormone Therapy  \n",
       "3    Hormone Therapy  \n",
       "4    Hormone Therapy  \n",
       "\n",
       "[5 rows x 71 columns]"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df =pd.read_csv('merged_df.csv')\n",
    "print(\"Merged DataFrame shape:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2dd10eac-4499-4bb9-ae55-872f11b64882",
   "metadata": {},
   "source": [
    "### Clean Data Frame\n",
    "#### Wraggle Function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "74ffd14b-6da5-4253-8746-07f7c00f8f6e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrangler(data):\n",
    "    \"\"\"\n",
    "    Drops rows with missing MEASURE_OF_RESPONSE orPATIENT_ID\n",
    "    Returns cleaned Dataframe.\n",
    "    \"\"\"\n",
    "    data =data.dropna(subset=['MEASURE_OF_RESPONSE'])\n",
    "    #dropping the patient_id colummn \n",
    "    data = data.drop(columns=['PATIENT_ID'])\n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "2a21799b-f32a-4d55-a0cc-f74efc968a57",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cleanedd DataFrame shape: (131, 70)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ARAF</th>\n",
       "      <th>BRAF</th>\n",
       "      <th>HRAS</th>\n",
       "      <th>KRAS</th>\n",
       "      <th>NRAS</th>\n",
       "      <th>RAF1</th>\n",
       "      <th>MAP2K1</th>\n",
       "      <th>MAP2K2</th>\n",
       "      <th>MAPK1</th>\n",
       "      <th>MAPK3</th>\n",
       "      <th>...</th>\n",
       "      <th>MAP3K1_cna</th>\n",
       "      <th>MAP3K2_cna</th>\n",
       "      <th>DUSP4_cna</th>\n",
       "      <th>DUSP6_cna</th>\n",
       "      <th>RASA1_cna</th>\n",
       "      <th>SPRY2_cna</th>\n",
       "      <th>FGFR1_cna</th>\n",
       "      <th>FGFR2_cna</th>\n",
       "      <th>MEASURE_OF_RESPONSE</th>\n",
       "      <th>TREATMENT_TYPE</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>-0.2260</td>\n",
       "      <td>1.1105</td>\n",
       "      <td>0.1345</td>\n",
       "      <td>-1.3060</td>\n",
       "      <td>0.4124</td>\n",
       "      <td>2.8254</td>\n",
       "      <td>1.2351</td>\n",
       "      <td>-0.4458</td>\n",
       "      <td>-0.4918</td>\n",
       "      <td>1.5008</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.036</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>stable disease</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>-0.2260</td>\n",
       "      <td>1.1105</td>\n",
       "      <td>0.1345</td>\n",
       "      <td>-1.3060</td>\n",
       "      <td>0.4124</td>\n",
       "      <td>2.8254</td>\n",
       "      <td>1.2351</td>\n",
       "      <td>-0.4458</td>\n",
       "      <td>-0.4918</td>\n",
       "      <td>1.5008</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.036</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.285</td>\n",
       "      <td>0.070</td>\n",
       "      <td>-1.070</td>\n",
       "      <td>-0.247</td>\n",
       "      <td>stable disease</td>\n",
       "      <td>Radiation Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.0553</td>\n",
       "      <td>-0.1687</td>\n",
       "      <td>-0.5289</td>\n",
       "      <td>-0.0316</td>\n",
       "      <td>-0.0179</td>\n",
       "      <td>0.8442</td>\n",
       "      <td>0.9456</td>\n",
       "      <td>-0.8248</td>\n",
       "      <td>-0.8615</td>\n",
       "      <td>-0.9742</td>\n",
       "      <td>...</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.006</td>\n",
       "      <td>0.004</td>\n",
       "      <td>-0.002</td>\n",
       "      <td>0.000</td>\n",
       "      <td>-0.701</td>\n",
       "      <td>0.004</td>\n",
       "      <td>-0.000</td>\n",
       "      <td>complete response</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>-1.5351</td>\n",
       "      <td>1.9882</td>\n",
       "      <td>-1.4200</td>\n",
       "      <td>1.5574</td>\n",
       "      <td>1.3370</td>\n",
       "      <td>-0.4445</td>\n",
       "      <td>-0.0737</td>\n",
       "      <td>-3.2238</td>\n",
       "      <td>0.5970</td>\n",
       "      <td>-1.4344</td>\n",
       "      <td>...</td>\n",
       "      <td>-0.006</td>\n",
       "      <td>-0.805</td>\n",
       "      <td>-0.804</td>\n",
       "      <td>-0.007</td>\n",
       "      <td>0.006</td>\n",
       "      <td>-0.737</td>\n",
       "      <td>2.150</td>\n",
       "      <td>0.728</td>\n",
       "      <td>clinical progressive disease</td>\n",
       "      <td>Hormone Therapy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>-1.1922</td>\n",
       "      <td>1.1519</td>\n",
       "      <td>-1.3305</td>\n",
       "      <td>2.6857</td>\n",
       "      <td>1.5874</td>\n",
       "      <td>-0.4955</td>\n",
       "      <td>0.3820</td>\n",
       "      <td>-1.9529</td>\n",
       "      <td>1.6209</td>\n",
       "      <td>-1.6199</td>\n",
       "      <td>...</td>\n",
       "      <td>0.044</td>\n",
       "      <td>-0.020</td>\n",
       "      <td>-0.041</td>\n",
       "      <td>0.004</td>\n",
       "      <td>0.005</td>\n",
       "      <td>-0.393</td>\n",
       "      <td>-0.027</td>\n",
       "      <td>-0.108</td>\n",
       "      <td>radiographic progressive disease</td>\n",
       "      <td>Radiation Therapy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 70 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      ARAF    BRAF    HRAS    KRAS    NRAS    RAF1  MAP2K1  MAP2K2   MAPK1  \\\n",
       "0  -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351 -0.4458 -0.4918   \n",
       "1  -0.2260  1.1105  0.1345 -1.3060  0.4124  2.8254  1.2351 -0.4458 -0.4918   \n",
       "3  -0.0553 -0.1687 -0.5289 -0.0316 -0.0179  0.8442  0.9456 -0.8248 -0.8615   \n",
       "8  -1.5351  1.9882 -1.4200  1.5574  1.3370 -0.4445 -0.0737 -3.2238  0.5970   \n",
       "10 -1.1922  1.1519 -1.3305  2.6857  1.5874 -0.4955  0.3820 -1.9529  1.6209   \n",
       "\n",
       "     MAPK3  ...  MAP3K1_cna  MAP3K2_cna  DUSP4_cna  DUSP6_cna  RASA1_cna  \\\n",
       "0   1.5008  ...      -0.285       0.036     -1.070      0.000     -0.285   \n",
       "1   1.5008  ...      -0.285       0.036     -1.070      0.000     -0.285   \n",
       "3  -0.9742  ...       0.000      -0.006      0.004     -0.002      0.000   \n",
       "8  -1.4344  ...      -0.006      -0.805     -0.804     -0.007      0.006   \n",
       "10 -1.6199  ...       0.044      -0.020     -0.041      0.004      0.005   \n",
       "\n",
       "    SPRY2_cna  FGFR1_cna  FGFR2_cna               MEASURE_OF_RESPONSE  \\\n",
       "0       0.070     -1.070     -0.247                    stable disease   \n",
       "1       0.070     -1.070     -0.247                    stable disease   \n",
       "3      -0.701      0.004     -0.000                 complete response   \n",
       "8      -0.737      2.150      0.728      clinical progressive disease   \n",
       "10     -0.393     -0.027     -0.108  radiographic progressive disease   \n",
       "\n",
       "       TREATMENT_TYPE  \n",
       "0     Hormone Therapy  \n",
       "1   Radiation Therapy  \n",
       "3     Hormone Therapy  \n",
       "8     Hormone Therapy  \n",
       "10  Radiation Therapy  \n",
       "\n",
       "[5 rows x 70 columns]"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Clean the data using wrangler function \n",
    "df_clean = wrangler(df)\n",
    "\n",
    "#Display the shape of the clean Dataframe\n",
    "print(\"Cleanedd DataFrame shape:\", df_clean.shape)\n",
    "\n",
    "#View the first few rows \n",
    "df_clean.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "6ab63921-9e99-4239-8f40-0c30187775c3",
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'MEASURE_OF_RESPONSSE'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3805\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3804\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m-> 3805\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_engine\u001b[38;5;241m.\u001b[39mget_loc(casted_key)\n\u001b[0;32m   3806\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m err:\n",
      "File \u001b[1;32mindex.pyx:167\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mindex.pyx:196\u001b[0m, in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7081\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "File \u001b[1;32mpandas\\\\_libs\\\\hashtable_class_helper.pxi:7089\u001b[0m, in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[1;34m()\u001b[0m\n",
      "\u001b[1;31mKeyError\u001b[0m: 'MEASURE_OF_RESPONSSE'",
      "\nThe above exception was the direct cause of the following exception:\n",
      "\u001b[1;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[1;32mIn[55], line 2\u001b[0m\n\u001b[0;32m      1\u001b[0m \u001b[38;5;66;03m# Count the treatment response categories \u001b[39;00m\n\u001b[1;32m----> 2\u001b[0m counts \u001b[38;5;241m=\u001b[39m df_clean[\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mMEASURE_OF_RESPONSSE\u001b[39m\u001b[38;5;124m'\u001b[39m]\u001b[38;5;241m.\u001b[39mvalue_counts()\n\u001b[0;32m      3\u001b[0m total \u001b[38;5;241m=\u001b[39m counts\u001b[38;5;241m.\u001b[39msum()\n\u001b[0;32m      4\u001b[0m percentages \u001b[38;5;241m=\u001b[39m (counts \u001b[38;5;241m/\u001b[39m total \u001b[38;5;241m*\u001b[39m \u001b[38;5;241m100\u001b[39m)\u001b[38;5;241m.\u001b[39mround(\u001b[38;5;241m1\u001b[39m)\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\pandas\\core\\frame.py:4102\u001b[0m, in \u001b[0;36mDataFrame.__getitem__\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   4100\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mnlevels \u001b[38;5;241m>\u001b[39m \u001b[38;5;241m1\u001b[39m:\n\u001b[0;32m   4101\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_getitem_multilevel(key)\n\u001b[1;32m-> 4102\u001b[0m indexer \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mcolumns\u001b[38;5;241m.\u001b[39mget_loc(key)\n\u001b[0;32m   4103\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m is_integer(indexer):\n\u001b[0;32m   4104\u001b[0m     indexer \u001b[38;5;241m=\u001b[39m [indexer]\n",
      "File \u001b[1;32m~\\anaconda3\\Lib\\site-packages\\pandas\\core\\indexes\\base.py:3812\u001b[0m, in \u001b[0;36mIndex.get_loc\u001b[1;34m(self, key)\u001b[0m\n\u001b[0;32m   3807\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28misinstance\u001b[39m(casted_key, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;129;01mor\u001b[39;00m (\n\u001b[0;32m   3808\u001b[0m         \u001b[38;5;28misinstance\u001b[39m(casted_key, abc\u001b[38;5;241m.\u001b[39mIterable)\n\u001b[0;32m   3809\u001b[0m         \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28many\u001b[39m(\u001b[38;5;28misinstance\u001b[39m(x, \u001b[38;5;28mslice\u001b[39m) \u001b[38;5;28;01mfor\u001b[39;00m x \u001b[38;5;129;01min\u001b[39;00m casted_key)\n\u001b[0;32m   3810\u001b[0m     ):\n\u001b[0;32m   3811\u001b[0m         \u001b[38;5;28;01mraise\u001b[39;00m InvalidIndexError(key)\n\u001b[1;32m-> 3812\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mKeyError\u001b[39;00m(key) \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01merr\u001b[39;00m\n\u001b[0;32m   3813\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m \u001b[38;5;167;01mTypeError\u001b[39;00m:\n\u001b[0;32m   3814\u001b[0m     \u001b[38;5;66;03m# If we have a listlike key, _check_indexing_error will raise\u001b[39;00m\n\u001b[0;32m   3815\u001b[0m     \u001b[38;5;66;03m#  InvalidIndexError. Otherwise we fall through and re-raise\u001b[39;00m\n\u001b[0;32m   3816\u001b[0m     \u001b[38;5;66;03m#  the TypeError.\u001b[39;00m\n\u001b[0;32m   3817\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_check_indexing_error(key)\n",
      "\u001b[1;31mKeyError\u001b[0m: 'MEASURE_OF_RESPONSSE'"
     ]
    }
   ],
   "source": [
    "# Count the treatment response categories \n",
    "counts = df_clean['MEASURE_OF_RESPONSSE'].value_counts()\n",
    "total = counts.sum()\n",
    "percentages = (counts / total * 100).round(1)\n",
    "\n",
    "#Create the bar plot\n",
    "colors = sns.color_palette('Dark2')\n",
    "ax = counts.plot(kind = 'barh', color=colors)\n",
    "\n",
    "#Add percentage labels next to each bar \n",
    "for i, (value, percent) in enumerate(zip(counts, percentages)):\n",
    "    plt.text(value + 1, i, f\"{percent}%\", va='center')\n",
    "\n",
    "#clean up plot\n",
    "ax.spines [['top', 'right']].set_visible(False)\n",
    "plt.xlabel(\"Number of Patients\")\n",
    "plt.title(\"Treatment Response Distribution\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "310aae74-0f6b-4bf0-95da-75f9330de2df",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Count the treatment response categories \n",
    "counts = df_clean['TREATMENT_TYPE'].value_counts()\n",
    "total = counts.sum()\n",
    "percentages = (counts / total * 100).round(1)\n",
    "\n",
    "#Create the bar plot\n",
    "colors = sns.color_palette('Dark2')\n",
    "ax = counts.plot(kind = 'barh', color=colors)\n",
    "\n",
    "#Add percentage labels next to each bar \n",
    "for i, (value, percent) in enumerate(zip(counts, percentages)):\n",
    "    plt.text(value + 1, i, f\"{percent}%\", va='center')\n",
    "\n",
    "#clean up plot\n",
    "ax.spines [['top', 'right']].set_visible(False)\n",
    "plt.xlabel(\"Number of Patients\")\n",
    "plt.title(\"Treatment Response Distribution\")\n",
    "plt.tight_layout()\n",
    "plt.show()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:base] *",
   "language": "python",
   "name": "conda-base-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
